<a href="https://colab.research.google.com/github/VitorFRodrigues/Alura_Challenges/blob/main/notebook/Alura_Challenge_DataScience1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
dados_URL = 'https://raw.githubusercontent.com/sthemonica/alura-voz/main/Dados/Telco-Customer-Churn.json'

In [2]:
import pandas as pd
dados = pd.read_json(dados_URL)

In [3]:
dados

customerID Churn                                           customer  \
0     0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1     0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2     0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3     0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4     0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
...          ...   ...                                                ...   
7262  9987-LUTYD    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
7263  9992-RRAMN   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7264  9992-UJOEL    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7265  9993-LHIEB    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7266  9995-HOTOH    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   

                                                  phone  \
0        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1       {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
...                                                 ...   
7262     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7263    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
7264     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7265     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7266  {'PhoneService': 'No', 'MultipleLines': 'No ph...   

                                               internet  \
0     {'InternetService': 'DSL', 'OnlineSecurity': '...   
1     {'InternetService': 'DSL', 'OnlineSecurity': '...   
2     {'InternetService': 'Fiber optic', 'OnlineSecu...   
3     {'InternetService': 'Fiber optic', 'OnlineSecu...   
4     {'InternetService': 'Fiber optic', 'OnlineSecu...   
...                                                 ...   
7262  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7263  {'InternetService': 'Fiber optic', 'OnlineSecu...   
7264  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7265  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7266  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                                account  
0     {'Contract': 'One year', 'PaperlessBilling': '...  
1     {'Contract': 'Month-to-month', 'PaperlessBilli...  
2     {'Contract': 'Month-to-month', 'PaperlessBilli...  
3     {'Contract': 'Month-to-month', 'PaperlessBilli...  
4     {'Contract': 'Month-to-month', 'PaperlessBilli...  
...                                                 ...  
7262  {'Contract': 'One year', 'PaperlessBilling': '...  
7263  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7264  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7265  {'Contract': 'Two year', 'PaperlessBilling': '...  
7266  {'Contract': 'Two year', 'PaperlessBilling': '...  

[7267 rows x 6 columns]

#1.0 Analise dos dados coluna a coluna 

##1.1 CustomerID

Checagem se cada ID é único

In [4]:
len(dados['customerID'].unique())

7267

Valores únicos de customerID é igual ao número de linhas dos dados o que comprova que cada linha possui ID único.

##1.2 Churn

Checagem se opção de escolha se limita a Yes / No

In [5]:
dados['Churn'].unique()

array(['No', 'Yes', ''], dtype=object)

Vemos que algumas linhas não possuem tag, vamos identificar quantas linhas:

In [10]:
len(dados.query('Churn!="Yes" & Churn!="No"'))

224

In [14]:
len(dados.query('Churn!="Yes" & Churn!="No"'))/len(dados)*100

3.0824274115866244

São 224 linhas sem Tag o que corresponde a cerca de 3% de todo conjunto de dados. Estas serão armazenada em variável específica e removida do conjunto de dados

In [16]:
dados_sem_tag = dados.query('Churn!="Yes" & Churn!="No"')
dados_sem_tag

customerID Churn                                           customer  \
30    0047-ZHDTW        {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
75    0120-YZLQA        {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
96    0154-QYHJU        {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
98    0162-RZGMZ        {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
175   0274-VVQOQ        {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
...          ...   ...                                                ...   
7158  9840-GSRFX        {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
7180  9872-RZQQB        {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
7211  9920-GNDMB        {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7239  9955-RVWSC        {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
7247  9966-VYRTZ        {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   

                                                  phone  \
30      {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
75       {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
96       {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
98       {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
175     {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
...                                                 ...   
7158    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
7180  {'PhoneService': 'No', 'MultipleLines': 'No ph...   
7211    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
7239     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7247     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                               internet  \
30    {'InternetService': 'Fiber optic', 'OnlineSecu...   
75    {'InternetService': 'No', 'OnlineSecurity': 'N...   
96    {'InternetService': 'DSL', 'OnlineSecurity': '...   
98    {'InternetService': 'DSL', 'OnlineSecurity': '...   
175   {'InternetService': 'Fiber optic', 'OnlineSecu...   
...                                                 ...   
7158  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7180  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7211  {'InternetService': 'Fiber optic', 'OnlineSecu...   
7239  {'InternetService': 'No', 'OnlineSecurity': 'N...   
7247  {'InternetService': 'No', 'OnlineSecurity': 'N...   

                                                account  
30    {'Contract': 'Month-to-month', 'PaperlessBilli...  
75    {'Contract': 'Two year', 'PaperlessBilling': '...  
96    {'Contract': 'One year', 'PaperlessBilling': '...  
98    {'Contract': 'Month-to-month', 'PaperlessBilli...  
175   {'Contract': 'One year', 'PaperlessBilling': '...  
...                                                 ...  
7158  {'Contract': 'One year', 'PaperlessBilling': '...  
7180  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7211  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7239  {'Contract': 'Two year', 'PaperlessBilling': '...  
7247  {'Contract': 'Month-to-month', 'PaperlessBilli...  

[224 rows x 6 columns]

In [22]:
dados_limpos = dados.query('Churn=="Yes" or Churn=="No"')
dados_limpos

customerID Churn                                           customer  \
0     0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1     0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2     0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3     0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4     0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
...          ...   ...                                                ...   
7262  9987-LUTYD    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
7263  9992-RRAMN   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7264  9992-UJOEL    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7265  9993-LHIEB    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7266  9995-HOTOH    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   

                                                  phone  \
0        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1       {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
...                                                 ...   
7262     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7263    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
7264     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7265     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7266  {'PhoneService': 'No', 'MultipleLines': 'No ph...   

                                               internet  \
0     {'InternetService': 'DSL', 'OnlineSecurity': '...   
1     {'InternetService': 'DSL', 'OnlineSecurity': '...   
2     {'InternetService': 'Fiber optic', 'OnlineSecu...   
3     {'InternetService': 'Fiber optic', 'OnlineSecu...   
4     {'InternetService': 'Fiber optic', 'OnlineSecu...   
...                                                 ...   
7262  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7263  {'InternetService': 'Fiber optic', 'OnlineSecu...   
7264  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7265  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7266  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                                account  
0     {'Contract': 'One year', 'PaperlessBilling': '...  
1     {'Contract': 'Month-to-month', 'PaperlessBilli...  
2     {'Contract': 'Month-to-month', 'PaperlessBilli...  
3     {'Contract': 'Month-to-month', 'PaperlessBilli...  
4     {'Contract': 'Month-to-month', 'PaperlessBilli...  
...                                                 ...  
7262  {'Contract': 'One year', 'PaperlessBilling': '...  
7263  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7264  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7265  {'Contract': 'Two year', 'PaperlessBilling': '...  
7266  {'Contract': 'Two year', 'PaperlessBilling': '...  

[7043 rows x 6 columns]

##1.3 Customer

Customer possui dados a serem normatizados

In [23]:
dados_limpos['customer'][0]

{'Dependents': 'Yes',
 'Partner': 'Yes',
 'SeniorCitizen': 0,
 'gender': 'Female',
 'tenure': 9}

Primeiramente vamos checar se cada linha possui dicionario com as mesmas keys

In [27]:
chave = []
for i in range(0, len(dados_limpos['customer'])):
  for key, value in dados_limpos['customer'][i].items():
    chave.append(key)

KeyError: ignored